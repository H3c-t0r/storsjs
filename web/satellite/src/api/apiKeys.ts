// Copyright (C) 2018 Storj Labs, Inc.
// See LICENSE for copying information.

import apollo from '@/utils/apolloManager';
import gql from 'graphql-tag';

export async function fetchAPIKeys(projectID: string) {
    let result: RequestResponse<any[]> = {
        errorMessage: '',
        isSuccess: false,
        data: []
    };

    try {
        let response: any = await apollo.query({
            query: gql(
                `query {
                    project(
                        id: "${projectID}",
                    ) {
                        apiKeys {
                            id,
                            name,
                            createdAt
                        }
                    }
                }`
            ),
            fetchPolicy: 'no-cache'
        });

        if (response.errors) {
            result.errorMessage = response.errors[0].message;
        } else {
            result.isSuccess = true;
            result.data = response.data.project.apiKeys;
        }
    } catch (e) {
        result.errorMessage = e.message;
    }

    return result;
}

export async function createAPIKey(projectID: string, name: string) {
    let result: RequestResponse<any> = {
        errorMessage: '',
        isSuccess: false,
        data: null
    };

    try {
        let response: any = await apollo.mutate({
            mutation: gql(
                `mutation {
                    createAPIKey(
                        projectID: "${projectID}",
                        name: "${name}"
                    ) {
                        key,
                        keyInfo {
                            id,
                            name,
                            createdAt
                        }
                    }
                }`
            ),
            fetchPolicy: 'no-cache'
        });

        if (response.errors) {
            result.errorMessage = response.errors[0].message;
        } else {
            console.log('response', response)
            result.isSuccess = true;
            result.data = {
                key: response.data.createAPIKey.key,
                keyInfo: response.data.createAPIKey.keyInfo
            }
        }
    } catch (e) {
        result.errorMessage = e.message;
    }

    return result;
}

export async function deleteAPIKey(id: string) {
    let result: RequestResponse<any> = {
        errorMessage: '',
        isSuccess: false,
        data: null
    };

    try {
        let response: any = await apollo.mutate({
            mutation: gql(
                `mutation {
                    deleteAPIKey(id: "${id}") {
                        id
                    }
                }`
            ),
            fetchPolicy: 'no-cache'
        });

        if (response.errors) {
            result.errorMessage = response.errors[0].message;
        } else {
            result.isSuccess = true;
            result.data = response.data.deleteAPIKey;
        }
    } catch (e) {
        result.errorMessage = e.message;
    }

    return result;
}