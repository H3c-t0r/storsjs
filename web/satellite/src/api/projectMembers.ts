// Copyright (C) 2018 Storj Labs, Inc.
// See LICENSE for copying information.

import apollo from '@/utils/apolloManager';
import gql from 'graphql-tag';

// Performs graqhQL request.
// Throws an exception if error occurs
export async function addProjectMember(userID: string, projectID: string): Promise<any> {
	let response: any = null;
	try {
		response = await apollo.mutate(
			{
				mutation: gql(`
                mutation {
                    addProjectMember(
                        projectID: "${projectID}",
                        userID: "${userID}"
                    ) {id}
                }`
				),
				fetchPolicy: 'no-cache',
			}
		);
	} catch (e) {
		// TODO: replace with popup in future
		console.error(e);
	}

	return response;
}

// Performs graqhQL request.
// Throws an exception if error occurs
export async function deleteProjectMember(userID: string, projectID: string): Promise<any> {
	let response: any = null;
	try {
		response = await apollo.mutate(
			{
				mutation: gql(`
                mutation {
                    deleteProjectMember(
                        projectID: "${projectID}",
                        userID: "${userID}"
                    ) {id}
                }`
				),
				fetchPolicy: 'no-cache',
			}
		);
	} catch (e) {
		// TODO: replace with popup in future
		console.error(e);
	}

	return response;
}

// Performs graqhQL request.
// Throws an exception if error occurs
export async function fetchProjectMembers(projectID: string, limit: string, offset: string): Promise<any> {
	let response: any = null;
	try {
		response = await apollo.query(
			{
				query: gql(`
                query {
                    project(
                        id: "${projectID}",
                    ) {
                        members(limit: ${limit}, offset: ${offset}) {
                            user {
                                id,
                                firstName,
                                lastName,
                                email,
                            },
                            joinedAt
                        }
                    }
                }`
				),
				fetchPolicy: 'no-cache',
			}
		);
	} catch (e) {
		// TODO: replace with popup in future
		console.error(e);
	}

	return response;
}
