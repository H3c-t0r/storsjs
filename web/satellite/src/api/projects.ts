// Copyright (C) 2018 Storj Labs, Inc.
// See LICENSE for copying information.

import apollo from '@/utils/apolloManager';
import gql from 'graphql-tag';

// Performs graqhQL request.
// Throws an exception if error occurs
export async function createProject(project: Project): Promise<any> {
	let response = await apollo.mutate(
		{
			mutation: gql(`
                mutation {
                    createProject(
                        input: {
                            name: "${project.name}",
                            description: "${project.description}",
                            companyName: "${project.companyName}",
                            isTermsAccepted: ${project.isTermsAccepted},
                        }
                    ) {id}
                }`
			),
			fetchPolicy: 'no-cache',
		}
	);

	if (!response) {
		// TODO: replace with popup in future
		console.error('cannot create project');

		return null;
	}

	return response;
}

// Performs graqhQL request for fetching all projects of current user.
export async function fetchProjects(): Promise<any> {
	let response = await apollo.query(
		{
			query: gql(`
                query {
                    myProjects{
                        name
                        id
                        description
                        createdAt            
                        ownerName           
                        companyName           
                        isTermsAccepted
                    }
                }`
			),
			fetchPolicy: 'no-cache',
		}
	);

	if (!response) {
		// TODO: replace with popup in future
		console.error('cannot fetch projects');

		return null;
	}

	return response;
}

// Performs graqhQL request for updating selected project description
export async function updateProject(projectID: string, description: string): Promise<any> {
	let response = await apollo.mutate(
		{
			mutation: gql(`
                mutation {
                    updateProjectDescription(
                        id: "${projectID}",
                        description: "${description}"
                    )
                }`
			),
			fetchPolicy: 'no-cache',
		}
	);

	if (!response) {
		// TODO: replace with popup in future
		console.error('cannot update project');

		return null;
	}

	return response;
}

// Performs graqhQL request for deleting selected project
export async function deleteProject(projectID: string): Promise<any> {
	let response = await apollo.mutate(
		{
			mutation: gql(`
                mutation {
                    deleteProject(
                        id: "${projectID}"
                    )
                }`
			),
			fetchPolicy: 'no-cache',
		}
	);

	if (!response) {
		// TODO: replace with popup in future
		console.error('cannot delete project');

		return null;
	}

	return response;
}
