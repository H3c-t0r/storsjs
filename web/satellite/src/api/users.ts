// Copyright (C) 2018 Storj Labs, Inc.
// See LICENSE for copying information.

import apolloManager from '../utils/apolloManager';
import gql from 'graphql-tag';

// Performs update user info graphQL mutation request.
// Returns User object if succeed, null otherwise
export async function updateBasicUserInfoRequest(user: User) {
    let response: any = null;

    try {
        response = await apolloManager.mutate(
            {
                mutation: gql(`
                    mutation {
                        updateUser (
                            id: "${user.id}",
                            input: {
                                email: "${user.email}",
                                firstName: "${user.firstName}",
                                lastName: "${user.lastName}"
                            }
                        ) {
                            email,
                            firstName,
                            lastName
                        }
                    }`
                ),
                fetchPolicy: 'no-cache',
            }
        );
    } catch (e) {
        console.error(e);
    }

	return response;
}

// Performs change password graphQL mutation
// Returns base user fields
export async function updatePasswordRequest(userId: string, password: string) {
    let response: any = null;

    try {
        response = await apolloManager.mutate(
            {
                mutation: gql(`
                    mutation {
                        updateUser (
                            id: "${userId}",
                            input: {
                                password: "${password}"
                            }
                        ) {
                            email,
                            firstName,
                            lastName
                        }
                    }`
                ),
                fetchPolicy: 'no-cache',
            }
        );
    } catch (e) {
        console.error(e);
    }

	return response;
}

// Performs Create user graqhQL request.
// Throws an exception if error occurs
// Returns object with newly created user
export async function createUserRequest(user: User, password: string): Promise<any> {
    let response: any = null;

    try {
        response = await apolloManager.mutate(
            {
                mutation: gql(`
                    mutation {
                        createUser(
                            input:{
                                email: "${user.email}",
                                password: "${password}",
                                firstName: "${user.firstName}",
                                lastName: "${user.lastName}",
                            }
                        )
                    }`
                ),
                fetchPolicy: 'no-cache',
            }
        );
    } catch (e) {
        console.error(e);
    }

	return response;
}

// Performs graqhQL request.
// Returns Token, User objects.
// Throws an exception if error occurs
export async function getTokenRequest(email: string, password: string): Promise<any> {
    let response: any = null;

    try {
        response = await apolloManager.query(
            {
                query: gql(`
                    query {
                        token(email: "${email}",
                              password: "${password}") {
                                  token
                        }
                    }`
                ),
                fetchPolicy: 'no-cache',
            }
        );
    } catch (e) {
        console.error(e);
    }

    return response;
}

// Performs graqhQL request.
// Returns Token, User objects.
// Throws an exception if error occurs
export async function getUserRequest(): Promise<any> {
    let response: any = null;

    try {
        response = await apolloManager.query(
            {
                query: gql(`
                    query {
                        user {
                            firstName,
                            lastName,
                            email,
                        }
                    }`
                ),
                fetchPolicy: 'no-cache',
            }
        );
    } catch (e) {
        console.error(e);
    }

    return response;
}

// Performs graqhQL request.
// User object.
// Throws an exception if error occurs
export async function deleteUserAccountRequest(password: string): Promise<any> {
    let response: any = null;

    try {
        response = await apolloManager.mutate(
            {
                mutation: gql(`
                    mutation {
                        deleteUser(password: "${password}") {
                            id
                        }
                    }`
                ),
                fetchPolicy: 'no-cache'
            }
        );
    } catch (e) {
        console.error(e);
    }

    return response;
}
