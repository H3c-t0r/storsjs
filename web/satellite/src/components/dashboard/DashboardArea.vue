// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

<template>
    <div>
        <EmptyState
            main-title="Create your first new project"
            :image-source="emptyImage"
        />
    </div>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator';

import EmptyState from '@/components/common/EmptyStateArea.vue';

import { EMPTY_STATE_IMAGES } from '@/utils/constants/emptyStatesImages';

@Component({
    components: {
        EmptyState,
    },
})
export default class DashboardArea extends Vue {
    public emptyImage: string = EMPTY_STATE_IMAGES.PROJECT;
}
</script>
