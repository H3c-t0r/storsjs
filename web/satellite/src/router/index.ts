import Vue from 'vue';
import Router from 'vue-router';
import ROUTES from '@/utils/constants/routerConstants';
import Login from '@/views/Login.vue';
import Register from '@/views/Register.vue';

Vue.use(Router);

export default new Router({
	mode: 'history',
	routes: [
		{
			path: ROUTES.DEFAULT.path,
			name: ROUTES.DEFAULT.name,
			component: Login
		},
		{
			path: ROUTES.REGISTER.path,
			name: ROUTES.REGISTER.name,
			component: Register
		}
  	]
});
