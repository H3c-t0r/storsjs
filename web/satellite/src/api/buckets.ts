// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

import { BaseGql } from '@/api/baseGql';
import { BucketCursor, BucketPage, BucketsApi } from '@/types/buckets';

/**
 * BucketsApiGql is a graphql implementation of Buckets API.
 * Exposes all bucket-related functionality
 */
export class BucketsApiGql extends BaseGql implements BucketsApi {
    /**
     * Fetch buckets
     *
     * @returns BucketPage
     * @throws Error
     */
    public async get(projectId: string, before: Date, cursor: BucketCursor): Promise<BucketPage> {
        const query =
            `query($projectId: String!, $before: DateTime!, $limit: Int!, $search: String!, $page: Int!) {
                project(id: $projectId) {
                    bucketUsages(before: $before, cursor: {
                        limit: $limit, search: $search, page: $page
                    }) {
                        bucketUsages {
                            bucketName,
                            storage,
                            egress,
                            objectCount,
                            since,
                            before
                        },
                        search,
                        limit,
                        offset,
                        pageCount,
                        currentPage,
                        totalCount
                    }
                }
            }`;

        const variables = {
            projectId,
            before: before.toISOString(),
            limit: cursor.limit,
            search: cursor.search,
            page: cursor.page,
        };

        const response = await this.query(query, variables);

        return this.fromJson(response.data.project.bucketUsages);
    }

    private fromJson(bucketPage): BucketPage {
        return new BucketPage(bucketPage.bucketUsages, bucketPage.search, bucketPage.limit, bucketPage.offset, bucketPage.pageCount, bucketPage.currentPage, bucketPage.totalCount);
    }
}
