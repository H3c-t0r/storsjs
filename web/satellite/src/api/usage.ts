// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

import { BaseGql } from '@/api/baseGql';
import { ProjectUsage, UsageApi } from '@/types/usage';

/**
 * Exposes all project-usage-related functionality
 */
export class ProjectUsageApiGql extends BaseGql implements UsageApi {
    /**
     * Fetch usage
     *
     * @returns ProjectUsage
     * @throws Error
     */
    public async get(projectId: string, since: Date, before: Date): Promise<ProjectUsage> {
        const query = `
            query($projectId: String!, $since: DateTime!, $before: DateTime!) {
                project(id: $projectId) {
                    usage(since: $since, before: $before) {
                        storage,
                        egress,
                        objectCount,
                        since,
                        before
                    }
                }
            }`;

        const variables = {
            projectId,
            since,
            before,
        };

        const response = await this.query(query, variables);

        return this.fromJson(response);
    }

    private fromJson(response: any): ProjectUsage {
        const usage = response.data.project.usage;

        return new ProjectUsage(usage.storage, usage.egress, usage.objectCount, usage.since, usage.before);
    }
}
