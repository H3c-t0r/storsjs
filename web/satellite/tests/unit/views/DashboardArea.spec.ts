// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

import Vuex from 'vuex';

import router, { RouteConfig } from '@/router';
import { makeApiKeysModule } from '@/store/modules/apiKeys';
import { appStateModule } from '@/store/modules/appState';
import { makeBucketsModule } from '@/store/modules/buckets';
import { makeNotificationsModule } from '@/store/modules/notifications';
import { makeProjectMembersModule } from '@/store/modules/projectMembers';
import { makeProjectsModule } from '@/store/modules/projects';
import { makeUsageModule } from '@/store/modules/usage';
import { makeUsersModule } from '@/store/modules/users';
import { User } from '@/types/users';
import { AuthToken } from '@/utils/authToken';
import { APP_STATE_ACTIONS } from '@/utils/constants/actionNames';
import { AppState } from '@/utils/constants/appStateEnum';
import DashboardArea from '@/views/DashboardArea.vue';
import { createLocalVue, shallowMount } from '@vue/test-utils';

import { ApiKeysMock } from '../mock/api/apiKeys';
import { BucketsMock } from '../mock/api/buckets';
import { ProjectMembersApiMock } from '../mock/api/projectMembers';
import { ProjectsApiMock } from '../mock/api/projects';
import { ProjectUsageMock } from '../mock/api/usage';
import { UsersApiMock } from '../mock/api/users';

const localVue = createLocalVue();
localVue.use(Vuex);

const usersApi = new UsersApiMock();
const projectsApi = new ProjectsApiMock();

usersApi.setMockUser(new User('1', '2', '3', '4', '5'));
projectsApi.setMockProjects([]);

const usersModule = makeUsersModule(usersApi);
const projectsModule = makeProjectsModule(projectsApi);
const apiKeysModule = makeApiKeysModule(new ApiKeysMock());
const teamMembersModule = makeProjectMembersModule(new ProjectMembersApiMock());
const bucketsModule = makeBucketsModule(new BucketsMock());
const usageModule = makeUsageModule(new ProjectUsageMock());
const notificationsModule = makeNotificationsModule();

const store = new Vuex.Store({
    modules: {
        notificationsModule,
        usageModule,
        bucketsModule,
        apiKeysModule,
        usersModule,
        projectsModule,
        appStateModule,
        teamMembersModule,
    },
});

describe('Dashboard', () => {
    beforeEach(() => {
        jest.resetAllMocks();
    });

    it('renders correctly when data is loading', () => {
        const wrapper = shallowMount(DashboardArea, {
            store,
            localVue,
            router,
        });

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.findAll('.loading-overlay.active').length).toBe(1);
        expect(wrapper.findAll('.dashboard-container__wrap').length).toBe(0);
    });

    it('renders correctly when data is loaded', () => {
        store.dispatch(APP_STATE_ACTIONS.CHANGE_STATE, AppState.LOADED);

        const wrapper = shallowMount(DashboardArea, {
            store,
            localVue,
            router,
        });

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.findAll('.loading-overlay active').length).toBe(0);
        expect(wrapper.findAll('.dashboard-container__wrap').length).toBe(1);
    });

    it('loads routes correctly when authorithed without project with available routes', async () => {
        jest.spyOn(AuthToken, 'get').mockReturnValue('authToken');

        const availableWithoutProject = [
            RouteConfig.Account.with(RouteConfig.Billing).path,
            RouteConfig.Account.with(RouteConfig.Profile).path,
        ];

        for (let i = 0; i < availableWithoutProject.length; i++) {
            const wrapper = await shallowMount(DashboardArea, {
                localVue,
                router,
                store,
            });

            setTimeout(() => {
                expect(wrapper.vm.$router.currentRoute.path).toBe(availableWithoutProject[i]);
            }, 50);
        }
    });

    it('loads routes correctly when authorithed without project with unavailable routes', async () => {
        jest.spyOn(AuthToken, 'get').mockReturnValue('authToken');

        const unavailableWithoutProject = [
            RouteConfig.ApiKeys.path,
            RouteConfig.Buckets.path,
            RouteConfig.Team.path,
            RouteConfig.ProjectOverview.with(RouteConfig.UsageReport).path,
        ];

        for (let i = 0; i < unavailableWithoutProject.length; i++) {
            await router.push(unavailableWithoutProject[i]);

            const wrapper = await shallowMount(DashboardArea, {
                localVue,
                router,
                store,
            });

            setTimeout(() => {
                expect(wrapper.vm.$router.currentRoute.path).toBe(RouteConfig.ProjectOverview.with(RouteConfig.ProjectDetails).path);
            }, 50);
        }

    });

    it('loads routes correctly when not authorithed', () => {
        const wrapper = shallowMount(DashboardArea, {
            store,
            localVue,
            router,
        });

        setTimeout(() => {
            expect(wrapper.vm.$router.currentRoute.path).toBe(RouteConfig.Login.path);
        }, 50);
    });
});
