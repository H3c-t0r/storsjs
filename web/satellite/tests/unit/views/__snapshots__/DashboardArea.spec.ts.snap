// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard renders correctly when data is loaded 1`] = `
<div class="dashboard-container">
  <!---->
  <div class="dashboard-container__wrap">
    <navigationarea-stub class="regular-navigation"></navigationarea-stub>
    <div class="dashboard-container__wrap__column">
      <dashboardheader-stub></dashboardheader-stub>
      <div class="dashboard-container__main-area">
        <vbanner-stub text="You have no payment method added." additionaltext="To start work with your account please add Credit Card or add $50.00 or more worth of STORJ tokens to your balance."></vbanner-stub>
        <div class="dashboard-container__main-area__content">
          <router-view-stub name="default"></router-view-stub>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Dashboard renders correctly when data is loading 1`] = `
<div class="dashboard-container">
  <div class="loading-overlay active"><img src="@/../static/images/register/Loading.gif" alt="Company logo loading gif" class="loading-image"></div>
  <!---->
</div>
`;
